### DISCUSSION TITLE
Job output contains a secret

### DISCUSSION THREAD

**Discussion by cerna** [community/community#27007] [Actions]
<p>Hello,</p>
<p>in my workflow I have a preparatory job that reads a settings file from repository and creates dynamic outputs from that. It creates a dynamic matrix for subsequent jobs, gets the <strong>ImageNameRoot</strong> which I use when creating Docker image name for uploading to Github Packages (the whole name then looks like <a href="http://docker.pkg.github.com/%24" class="inline-onebox-loading" rel="nofollow noopener">docker.pkg.github.com/$</a>{ <strong>user</strong> }/${ <strong>repository</strong> }/${ <strong>ImageNameRoot</strong> }${ <strong>ImageType</strong> }:${ <strong>Version</strong> }).</p>
<p>I also have in my repository <strong>secrets</strong> storage vault stored <strong>login</strong> and <strong>password</strong> to third party services. It just so happens that the login is subset of <strong>ImageNameRoot,</strong> i.e the login sequence of characters is wholly included in <strong>ImageNameRoot</strong> sequence of characters. (Let me reiterate that everybody who looks at the workflow/repository can find out what the <strong>ImageNameRoot</strong> is.)</p>
<p>Now I discovered, that during run of the preparatory job, there was a yellow windows (I don’t have it stored and as I re-ran the workflow, it disappeared) informing me that the job output can contain a <em>secret</em> and it was redacted. Problem is, the <strong>ImageNameRoot</strong> output become empty string. Given that I use this name as a Github Package name, it would create and upload a newly named package. (Fortunately, I was watching it and stopped it in time. It is a public repository, so I wouldn’t be able to delete this package.)</p>
<p>I consider it a <strong>pretty severe bug</strong>. What I would expect to happen is to either fail on the preparatory job stage when it cannot export an output or to actually ascertain that the output is not from <strong>secrets</strong> vault and do nothing, i.e. don’t redact. But in <strong>no way</strong> to output an empty string. (As it is now the Github Actions logic actually has negative output on security, as it tells everybody what is the content of given <strong>secret</strong>.)</p>
<p>So far I changed (deleted) the third party login/password service. But can I do anything else about it?</p>
<p>I cannot store the 3rd party login in repository, as it would clash against the fork+work+pull flow (every developer can/needs to have his own login/password combination in repository). The image’s names are the same for everybody, as these are just tester runners and are separated from each other by the ${ <strong>user</strong> }/${ <strong>repository</strong> } combination.</p>
<p>(The actual workflow is <a href="https://github.com/machinekit/machinekit-hal/blob/e81023b8b9ae49a9e971d9da6d424297061772bc/.github/workflows/debian-builder-workflow.yaml#L51" rel="nofollow noopener">this</a>.)</p>
<p>Thank you.</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/cerna">@cerna</a> ,</p>
<p>Thank you for reaching this out! I checked your workflow yaml, the <strong>ImageNameRoot </strong> output value comes from ‘./machinekit-hal/scripts/debian-distro-settings.json’, however not any secrets related. The output value is displayed on the printf command.</p>
<p><img src="https://user-images.githubusercontent.com/52772761/181101136-227f12b6-5c3f-4ada-8757-eb6a898c369b.png" alt="outputvalue.png" data-base62-sha1="bjubks7SYRjtyPdNDUPcwUjo8sy" width="288" height="93"></p>
<p>For ouput value, it will always display value in logs, it’s a limitation that cannot avoid data leaking with it.</p>
<p>If you include secrets in the output value, it’s by designed that empty will display, checked on my side as below:</p>
<p><img src="https://user-images.githubusercontent.com/52772761/181101137-527a7e7e-3bea-4672-93c8-8401a22c4e5f.png" alt="outputsecrets.png" data-base62-sha1="zzWboMQvA1iYZH7irTnrBDAharL" width="535" height="279"></p>
<p>Please still avoid printing secrets to the log intentionally.</p>
<p>Thanks.</p>
